### PR TITLE
enhancer update

### PR DESCRIPTION
**Pull request by Moeblack**
I have added a method to enhance the quality of faces using Codeformer. Currently, it is done automatically.

### CODE REVIEW

File: roop/core.py [hunk 1]
```diff
@@ -194,10 +195,14 @@ def start() -> None:
     extract_frames(roop.globals.target_path)
     temp_frame_paths = get_temp_frame_paths(roop.globals.target_path)
     update_status('Swapping in progress...')
-    conditional_process_video(roop.globals.source_path, temp_frame_paths)
-    # prevent memory leak using ffmpeg with cuda
+    conditional_process_video(roop.globals.source_path, temp_frame_paths, process_video)
```
Comment (henryruhs) [inline roop/core.py:198]: This should be `roop.enhancer.process_video` - change the imports

File: roop/core.py [hunk 2]
```diff
@@ -194,10 +195,14 @@ def start() -> None:
...
     if roop.globals.gpu_vendor == 'nvidia':
         torch.cuda.empty_cache()
+    update_status('enhancinging in progress...')
+    conditional_process_video(roop.globals.source_path, temp_frame_paths, roop.enhancer.process_video)
+    # prevent memory leak using ffmpeg with cuda
```
Comment (henryruhs) [inline roop/core.py:203]: I guess this can go to a method release_resources() that you put right under limit_resources() and we call it after swapper and enhancer once

File: roop/enhancer.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+import roop.globals
+
+from torchvision.transforms.functional import normalize
+
+from codeformer.facelib.utils.face_restoration_helper import FaceRestoreHelper
+from codeformer.basicsr.utils.download_util import load_file_from_url
+from codeformer.basicsr.utils.registry import ARCH_REGISTRY
+from codeformer.basicsr.utils import img2tensor, tensor2img
+
+pretrain_model_url = {
```
Comment (henryruhs) [inline roop/enhancer.py:16]: Why does this exist? It will download the weights automatically?

File: roop/enhancer.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    )
+
+#FACE_HELPER = None
+CODE_FORMER = None
+THREAD_LOCK = threading.Lock()
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+ckpt_path = "CodeFormer/weights/CodeFormer/codeformer.pth"
+checkpoint = torch.load(ckpt_path)["params_ema"]
+
+def get_facepaste_enhancer():
```
Comment (henryruhs) [inline roop/enhancer.py:48]: get_face_enhancer and FACE_ENHANCER as the instance variable

File: roop/enhancer.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    )
+if not os.path.exists("CodeFormer/weights/realesrgan/RealESRGAN_x2plus.pth"):
+    load_file_from_url(
+        url=pretrain_model_url["realesrgan"], model_dir="CodeFormer/weights/realesrgan", progress=True, file_name=None
+    )
+
+#FACE_HELPER = None
+CODE_FORMER = None
+THREAD_LOCK = threading.Lock()
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (henryruhs) [inline roop/enhancer.py:44]: All this calls here have to go - once torch is imported it breaks AMD

File: roop/enhancer.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        use_parse=True,
+        device=device,
+    )
+    return FACE_HELPER
+
+
+def enhance_face_in_frame(cropped_faces):
+    try:
+        for idx, cropped_face in enumerate(cropped_faces):
+            face_t = data_preprocess(cropped_face)
```
Comment (henryruhs) [inline roop/enhancer.py:81]: do not use idx, face_t ... write full understandable names

File: roop/utilities.py [hunk 7]
```diff
@@ -74,7 +74,8 @@ def create_temp(target_path: str) -> None:
 def move_temp(target_path: str, output_path: str) -> None:
     temp_output_path = get_temp_output_path(target_path)
     if os.path.isfile(temp_output_path):
-        shutil.move(temp_output_path, output_path)
+        # shit.move will not cover the file if it exists.
```
Comment (henryruhs) [inline roop/utilities.py]: Remove comment

File: roop/enhancer.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        face_helper.clean_all()
+        return result
+    except RuntimeError as error:
+        print(f"Failed inference for CodeFormer-code-paste: {error}")
+
+
+
+
+def data_preprocess(frame):
+    frame_t = img2tensor(frame / 255.0, bgr2rgb=True, float32=True)
```
Comment (henryruhs) [inline roop/enhancer.py:110]: what is frame_t ?

File: roop/enhancer.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+import cv2
+import numpy as np
+import torch
+import threading
+from tqdm import tqdm
+import roop.globals
+
+from torchvision.transforms.functional import normalize
+
+from codeformer.facelib.utils.face_restoration_helper import FaceRestoreHelper
```
Comment (henryruhs) [inline roop/enhancer.py:11]: There is nothing added to the requirements.txt?

File: roop/enhancer.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+                n_head=8,
+                n_layers=9,
+                connect_list=["32", "64", "128", "256"],
+            ).to(device)
+            CODE_FORMER.load_state_dict(checkpoint)
+            CODE_FORMER.eval()
+        return CODE_FORMER
+
+
+def get_facepaste_back(FACE_HELPER):
```
Comment (henryruhs) [inline roop/enhancer.py:64]: get_face_enhancer and FACE_ENHANCER as the instance variable

File: roop/enhancer.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    if FACE_HELPER is None:
+        FACE_HELPER = FaceRestoreHelper(
```
Comment (henryruhs) [inline roop/enhancer.py:66]: Why is there no THREAD_LOCK here?

File: roop/enhancer.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+        result = face_helper.paste_faces_to_input_image()
+        face_helper.clean_all()
+        return result
+    except RuntimeError as error:
+        print(f"Failed inference for CodeFormer-code-paste: {error}")
+
+
+def data_preprocess(frame):
+    frame_t = img2tensor(frame / 255.0, bgr2rgb=True, float32=True)
+    normalize(frame_t, (0.5, 0.5, 0.5), (0.5, 0.5, 0.5), inplace=True)
```
Comment (henryruhs) [inline roop/enhancer.py:111]: normalize what? please improve the method name

File: roop/core.py [hunk 13]
```diff
@@ -194,10 +195,11 @@ def start() -> None:
     extract_frames(roop.globals.target_path)
     temp_frame_paths = get_temp_frame_paths(roop.globals.target_path)
     update_status('Swapping in progress...')
-    conditional_process_video(roop.globals.source_path, temp_frame_paths)
-    # prevent memory leak using ffmpeg with cuda
+    conditional_process_video(roop.globals.source_path, temp_frame_paths, roop.swapper.process_video)
     if roop.globals.gpu_vendor == 'nvidia':
         torch.cuda.empty_cache()
+    update_status('enhancinging in progress...')
```
Comment (henryruhs) [inline roop/core.py:201]: Load conditional according to `--frame-processor`

core:
```
parser.add_argument('--frame-processor', help='list of frame processors to run', dest='frame-processor', default='face-swapper', choices=['face-swapper', 'face-enhancer'])
```

```
roop.globals.frame_processors = args.frame_processor
```
globals:
```
frame_processors = None
```

File: roop/core.py [hunk 14]
```diff
@@ -178,7 +179,7 @@ def start() -> None:
     if has_image_extention(roop.globals.target_path):
         if predict_image(roop.globals.target_path) > 0.85:
             destroy()
-        process_image(roop.globals.source_path, roop.globals.target_path, roop.globals.output_path)
+        roop.swapper.process_image(roop.globals.source_path, roop.globals.target_path, roop.globals.output_path)
```
Comment (henryruhs) [inline roop/core.py:182]: roop.enhancer.process_image is missing here

File: roop/enhancer.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+    return frame_t.unsqueeze(0).to(device)
+
+
+def generate_output(frame_t, codeformer_fidelity = 0.6):
```
Comment (henryruhs) [inline roop/enhancer.py:115]: method has bad naming - does not tell me what it does

File: roop/enhancer.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+    with torch.no_grad():
+        output = get_facepaste_enhancer()(frame_t, w=codeformer_fidelity, adain=True)[0]
+    return output
+
+
+def postprocess_output(output):
```
Comment (henryruhs) [inline roop/enhancer.py:121]: method has bad naming - does not tell me what it does

File: roop/enhancer.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+    restored_face = tensor2img(output, rgb2bgr=True, min_max=(-1, 1))
+    return restored_face.astype("uint8")
+
+
+def restore_face(face_t):
```
Comment (henryruhs) [inline roop/enhancer.py:126]: this try catch in restore_face and process_frames have to go somehow

File: roop/enhancer.py [hunk 18]
```diff
@@ -0,0 +1,185 @@
+import os
+import cv2
+import numpy as np
```
Comment (henryruhs) [inline roop/enhancer.py:3]: Unused import (see lint task in pipeline for more)

File: requirements.txt [hunk 19]
```diff
@@ -7,13 +7,14 @@ insightface==0.7.3
 psutil==5.9.5
 tk==0.1.0
 pillow==9.5.0
-torch==2.0.1+cu118; sys_platform != 'darwin'
-torch==2.0.1; sys_platform == 'darwin'
+torch==2.0.1+cu118
```
Comment (henryruhs) [inline requirements.txt:10]: Do not remove the conditional installation for darwin!

File: roop/core.py (hunk 14 above)
Comment (henryruhs) [reply roop/core.py:184]: this is not resolved... you have to run both processors according to the args

File: roop/core.py [hunk 20]
```diff
@@ -193,11 +196,17 @@ def start() -> None:
...
     temp_frame_paths = get_temp_frame_paths(roop.globals.target_path)
-    update_status('Swapping in progress...')
-    conditional_process_video(roop.globals.source_path, temp_frame_paths)
-    # prevent memory leak using ffmpeg with cuda
+    if 'face-swapper' in roop.globals.frame_processors:
+        update_status('Swapping in progress...')
+        conditional_process_video(roop.globals.source_path, temp_frame_paths, roop.swapper.process_video)
     if roop.globals.gpu_vendor == 'nvidia':
         torch.cuda.empty_cache()
+    if 'ROCMExecutionProvider' in roop.globals.providers:
```
Comment (henryruhs) [inline roop/core.py:204]: if this is nvidia/cuda only, then please do remove the ROCM check and just is `if roop.globals.cpu_vendor = 'nvidia'` instead

File: roop/core.py [hunk 21]
```diff
@@ -38,6 +39,7 @@ def parse_args() -> None:
     parser.add_argument('-f', '--face', help='use a face image', dest='source_path')
     parser.add_argument('-t', '--target', help='replace image or video with face', dest='target_path')
     parser.add_argument('-o', '--output', help='save output to this file', dest='output_path')
+    parser.add_argument('--frame-processor', help='list of frame processors to run', dest='frame-processor', default='face-swapper', choices=['face-swapper', 'face-enhancer'])
```
Comment (henryruhs) [inline roop/core.py]: dest='frame_processor' (underscore)
